mesonbuild/interpreter/mesonmain.py

```diff
@@ -294,21 +295,52 @@ def override_find_program_method(self, args, kwargs):
             raise InterpreterException('Second argument must be an external program or executable.')
         self.interpreter.add_find_program_override(name, exe)
 
+    @typed_kwargs('meson.override_dependency', NATIVE_KW,
+                  KwargInfo('static', bool, since='0.60.0'))
+    @typed_pos_args('meson.override_dependency', str, dependencies.Dependency)
     @FeatureNew('meson.override_dependency', '0.54.0')
-    @permittedKwargs({'native'})
     def override_dependency_method(self, args, kwargs):
-        if len(args) != 2:
-            raise InterpreterException('Override needs two arguments')
-        name = args[0]
-        dep = args[1]
-        if not isinstance(name, str) or not name:
+        name, dep = args
+        if not name:
             raise InterpreterException('First argument must be a string and cannot be empty')
-        if not isinstance(dep, dependencies.Dependency):
-            raise InterpreterException('Second argument must be a dependency object')
+
+        optkey = OptionKey('default_library', subproject=self.interpreter.subproject)
+        default_library = self.interpreter.coredata.get_option(optkey)
+        static = kwargs['static']
+        if static is None:
+            # We don't know if dep represents a static or shared library, could
+            # be a mix of both. We assume it is following default_library
+            # value.
+            self._override_dependency_impl(name, dep, kwargs, static=None)
+            if default_library == 'static':
+                self._override_dependency_impl(name, dep, kwargs, static=True)
+            elif default_library == 'shared':
+                self._override_dependency_impl(name, dep, kwargs, static=False)
+            else:
+                self._override_dependency_impl(name, dep, kwargs, static=True)
+                self._override_dependency_impl(name, dep, kwargs, static=False)
+        else:
+            # dependency('foo') without specifying static kwarg should find this
+            # override regardless of the static value here. But do not raise error
+            # if it has already been overridden, which would happend when overriding
+            # static and shared separately:
+            # meson.override_dependency('foo', shared_dep, static: false)
+            # meson.override_dependency('foo', static_dep, static: true)
+            # In that case dependency('foo') would return the first override.
+            self._override_dependency_impl(name, dep, kwargs, static=None, permissive=True)
+            self._override_dependency_impl(name, dep, kwargs, static=static)
+
+    def _override_dependency_impl(self, name, dep, kwargs, static, permissive=False):
+        if static is None:
+            kwargs.pop('static', None)
+        else:
+            kwargs['static'] = static
```

Comment:
This seems sketchy, you're mutating the state that's going to be re-used in the calling context. I feel like we should be copying here before mutating. Also, we know static is going to be in here, so just use `del`

Reply:
static is not always there because we could have removed it in a previous _override_dependency_impl() call. I changed now to always do a copy, in that case we are sure the `static` kwarg is always there and we can use del.